src/Http/Http/perf/Microbenchmarks/RequestTimeoutsMiddlewareBenchmark.cs

```diff
@@ -0,0 +1,148 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using BenchmarkDotNet.Attributes;
+using Microsoft.AspNetCore.Http.Timeouts;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Http;
+
+public class RequestTimeoutsMiddlewareBenchmark
+{
+    RequestTimeoutsMiddleware _middlewareWithNoTimeout;
+    RequestTimeoutsMiddleware _middleware;
+    RequestTimeoutsMiddleware _middlewareWithThrow;
+
+    [GlobalSetup]
+    public void GlobalSetup()
+    {
+        _middlewareWithNoTimeout = new RequestTimeoutsMiddleware(
+            context => Task.CompletedTask,
+            new CancellationTokenLinker(),
+            NullLogger<RequestTimeoutsMiddleware>.Instance,
+            Options.Create(new RequestTimeoutOptions()));
+
+        _middleware = new RequestTimeoutsMiddleware(
+          context => Task.CompletedTask,
+          new CancellationTokenLinker(),
+          NullLogger<RequestTimeoutsMiddleware>.Instance,
+          Options.Create(new RequestTimeoutOptions
+          {
+              DefaultPolicy = new RequestTimeoutPolicy
+              {
+                  Timeout = TimeSpan.FromMilliseconds(200)
+              },
+              Policies =
+              {
+                  ["policy1"] = new RequestTimeoutPolicy { Timeout = TimeSpan.FromMilliseconds(200)}
+              }
+          }));
+
+        _middlewareWithThrow = new RequestTimeoutsMiddleware(
+          async context =>
+          {
+              await Task.Delay(TimeSpan.FromMicroseconds(2));
+              context.RequestAborted.ThrowIfCancellationRequested();
+          },
+          new CancellationTokenLinker(),
+          NullLogger<RequestTimeoutsMiddleware>.Instance,
+          Options.Create(new RequestTimeoutOptions
+          {
+              DefaultPolicy = new RequestTimeoutPolicy
+              {
+                  Timeout = TimeSpan.FromMicroseconds(1)
+              }
+          }));
+    }
+
+    [Benchmark]
+    public async Task NoMetadataNoDefault()
+    {
+        var context = new DefaultHttpContext();
```

Comment:
DefaultHttpContext doesn't provided a cancellable RequestAborted token, you'll need to set one. That will change your allocations when it actually has to link the tokens.

Reply:
I've set the cancellation token, but the results seems like the previous one. Is there anything I'm missing?
|                           Method |       Mean |     Error |    StdDev |        Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------------------- |-----------:|----------:|----------:|------------:|-------:|------:|------:|----------:|
|              NoMetadataNoDefault |   754.6 ns |  12.26 ns |  10.86 ns | 1,325,161.4 | 0.0391 |     - |     - |      1 KB |
|                   DefaultTimeout | 1,790.9 ns |  27.08 ns |  25.33 ns |   558,380.6 | 0.0038 |     - |     - |      2 KB |
| DefaultTimeoutOverridenByDisable | 1,396.5 ns |   5.54 ns |   5.18 ns |   716,098.0 | 0.0744 |     - |     - |      2 KB |
|                  TimeoutMetadata | 2,854.4 ns |  56.73 ns | 149.46 ns |   350,336.6 | 0.0038 |     - |     - |      3 KB |
|              NamedPolicyMetadata | 2,801.9 ns |  56.03 ns | 139.54 ns |   356,906.3 | 0.0038 |     - |     - |      3 KB |
|                     TimeoutFires | 5,231.8 ns | 197.57 ns | 582.53 ns |   191,140.4 | 0.0534 |     - |     - |      2 KB |

Reply:
Great, that's better than expected. I recommended one more benchmark change to make the next delegate yield which should cause more state machines, but otherwise this seems good.